task_processing/plugins/mesos/execution_framework.py

```diff
@@ -15,338 +13,118 @@
-        # if total > 0 and failed_constraints == total:
-        #     raise ConstraintFailAllTasksError
-
-        return tasks, state
-
-    def offer_matches_pool(self, offer):
-        # for attribute in offer.attributes:
-            # if attribute.name == "pool":
-                # return attribute.text.value == self.service_config.get_pool()
-        # # we didn't find a pool attribute on this slave, so assume it's
-        # # not in our pool.
-        # return False
-        return True
-
-    def within_reconcile_backoff(self):
-        return time.time() - self.reconcile_backoff < self.reconcile_start_time
-
-    def kill_task(self, driver, task):
+    def kill_task(self, task_id):
         tid = mesos_pb2.TaskID()
-        tid.value = task
-        driver.killTask(tid)
+        tid.value = task_id
+        self.driver.killTask(tid)
 
     def blacklist_slave(self, slave_id):
         if slave_id in self.blacklisted_slaves:
-            return
-
-        log.debug("Blacklisting slave: %s" % slave_id)
-        with self.blacklisted_slaves_lock:
-            self.blacklisted_slaves.add(slave_id)
-            Timer(self.blacklist_timeout,
-                  lambda: self.unblacklist_slave(slave_id)).start()
-
-    def unblacklist_slave(self, slave_id):
-        if slave_id not in self.blacklisted_slaves:
-            return
-
-        log.debug("Unblacklisting slave: %s" % slave_id)
-        with self.blacklisted_slaves_lock:
-            self.blacklisted_slaves.discard(slave_id)
-
-    def enqueue(self, task_info):
-        self.tasks_waiting.put(task_info)
+            # Punish this slave for more time.
+            self.blacklisted_slaves.pop(slave_id, None)
 
-    def wait_for(self, task_id):
-        result = Queue()
-        self.on_success(task_id, lambda: result.put(0))
-        self.on_failure(task_id, lambda: result.put(1))
-        return result.get()
-
-    def on_success(self, task_id, callback):
-        pass
-
-    def on_failure(self, task_id, callback):
-        pass
+        log.info('Blacklisting slave: {id} for {secs} seconds.'.format(
+            id=slave_id,
+            secs=self.slave_blacklist_timeout_s
+        )
+        )
+        self.blacklisted_slaves[slave_id] = time.time()
 
-    def on_status(self, task_id, func):
-        pass
+    def unblacklist_slaves(self):
+        while True:
+            time_now = time.time()
+            for slave_id in self.blacklisted_slaves.keys():
+                if time_now < (
+                    self.blacklisted_slaves[slave_id] +
+                    self.slave_blacklist_timeout_s
+                ):
+                    log.info('Unblacklisting slave: {id}'.format(id=slave_id))
+                    self.blacklisted_slaves.pop(slave_id, None)
+            time.sleep(10)
+
+    def enqueue_task(self, task):
+        # TODO: Add a wrapper for this task
+        self.task_metadata[task.task_id] = {
```

Comment:
I am not sure I understand the need for PRecord. We need something mutable. We can even have a class:

class TaskMetada:
        def __init__(self, task_state, retries=0, slave_name, task_config):
            self.task_state = task_state
            self.retries = retries
            slave_name = slave_name
            task_config = task_config

Then task_metadata will look like:
self.task_metadata[task_id] = TaskMetadata(...)